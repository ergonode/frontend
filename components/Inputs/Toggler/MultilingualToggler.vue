/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div :class="['multilingual-toggler', {'multilingual-toggler--disabled': disabled}]">
        <Toggler
            :value="value"
            :disabled="disabled"
            @input="onValueChange" />
        <span
            class="multilingual-toggler__label font--medium-12-16"
            v-text="label" />
        <InfoHint :hint="hint" />
    </div>
</template>

<script>
import Toggler from '~/components/Inputs/Toggler/Toggler';
import InfoHint from '~/components/Inputs/Hint/InfoHint';

export default {
    name: 'MultilingualToggler',
    components: {
        Toggler,
        InfoHint,
    },
    props: {
        value: {
            type: Boolean,
            required: true,
        },
        disabled: {
            type: Boolean,
            default: false,
        },
        label: {
            type: String,
            default: '',
        },
        hint: {
            type: String,
            default: '',
        },
    },
    methods: {
        onValueChange() {
            this.$emit('input', !this.value);
        },
    },
};
</script>

<style lang="scss" scoped>
    .multilingual-toggler {
        $TOGGLER: &;

        display: flex;

        &--disabled {
            #{$TOGGLER}__label {
                color: $GRAPHITE_LIGHT;
            }
        }

        &__label {
            display: flex;
            align-items: center;
            margin: 0 4px 0 8px;
            color: $GRAPHITE_DARK;
        }
    }
</style>
