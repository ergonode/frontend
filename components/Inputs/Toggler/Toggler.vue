/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div
        :class="['toggler', togglerStateClasses]"
        @click="onClick">
        <div class="toggler__state" />
    </div>
</template>

<script>

export default {
    name: 'Toggler',
    props: {
        value: {
            type: Boolean,
            default: false,
        },
        disabled: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        togglerStateClasses() {
            return [{
                'toggler--selected': this.value,
                'toggler--disabled': this.disabled,
            }];
        },
    },
    methods: {
        onClick() {
            this.$emit('input', !this.value);
        },
    },
};
</script>

<style lang="scss" scoped>
    .toggler {
        $TOGGLER: &;

        position: relative;
        display: flex;
        justify-content: flex-start;
        align-items: center;
        width: 30px;

        &::before {
            position: absolute;
            top: 50%;
            left: 0;
            transform: translate(0, -50%);
            width: 100%;
            height: 12px;
            border-radius: 8px;
            box-shadow: $ELEVATOR_HOLE;
            content: "";
        }

        &__state {
            position: relative;
            z-index: 1;
            display: flex;
            width: 16px;
            height: 16px;
            border-radius: 8px;
            background-color: $WHITE;
            box-shadow: $ELEVATOR_2_DP;

            &::after {
                position: absolute;
                top: 5px;
                left: 4px;
                width: 6px;
                height: 3px;
                border-left: 2px solid $WHITE;
                border-bottom: 2px solid $WHITE;
                transform: rotate(-45deg);
                opacity: 0;
                content: "";
            }
        }

        &--selected {
            justify-content: flex-end;

            #{$TOGGLER}__state {
                background-color: $GREEN;
            }

            #{$TOGGLER}__state::after {
                opacity: 1;
            }
        }

        &--disabled {
            pointer-events: none;

            #{$TOGGLER}__state {
                background-color: $GREY_DARK;
            }
        }
    }
</style>
