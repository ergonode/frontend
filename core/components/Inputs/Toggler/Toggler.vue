/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="toggler">
        <input
            :id="associatedLabel"
            :value="value"
            :disabled="disabled"
            type="checkbox"
            @input="onValueChange">
        <label :for="associatedLabel">
            <div class="toggler__state" />
        </label>
        <span
            v-if="label"
            class="toggler__label"
            v-text="label" />
        <slot name="append" />
    </div>
</template>

<script>
export default {
    name: 'Toggler',
    props: {
        value: {
            type: Boolean,
            default: false,
        },
        disabled: {
            type: Boolean,
            default: false,
        },
        label: {
            type: String,
            default: '',
        },
    },
    data() {
        return {
            associatedLabel: '',
        };
    },
    mounted() {
        // eslint-disable-next-line no-underscore-dangle
        this.associatedLabel = `input-${this._uid}`;
    },
    methods: {
        onValueChange() {
            this.$emit('input', !this.value);
        },
    },
};
</script>

<style lang="scss" scoped>
    .toggler {
        $toggler: &;

        position: relative;
        display: grid;
        grid-template-columns: max-content max-content;
        grid-auto-flow: column;
        column-gap: 8px;
        align-items: center;

        & label {
            position: relative;
            display: flex;
            flex: 1;
            align-items: center;
            width: 30px;

            &::before {
                position: absolute;
                top: 50%;
                left: 0;
                transform: translate(0, -50%);
                width: 100%;
                height: 12px;
                background-color: $GREY_LIGHT;
                border-radius: 8px;
                box-shadow: $ELEVATOR_HOLE;
                content: "";
            }
        }

        &__state {
            position: relative;
            z-index: 1;
            display: flex;
            width: 16px;
            height: 16px;
            border-radius: 8px;
            background-color: $WHITE;
            box-shadow: $ELEVATOR_2_DP;
            transition: all 0.3s cubic-bezier(0.25, 0.8, 0.5, 1);

            &::after {
                position: absolute;
                top: 5px;
                left: 4px;
                width: 6px;
                height: 3px;
                border-left: 2px solid $WHITE;
                border-bottom: 2px solid $WHITE;
                transform: rotate(-45deg);
                opacity: 0;
                content: "";
            }
        }

        &__label {
            color: $GRAPHITE_DARK;
            font: $FONT_MEDIUM_12_16;
        }

        & input[type="checkbox"] {
            position: absolute;
            top: 0;
            left: 0;
            visibility: hidden;

            &:checked + label {
                #{$toggler}__state {
                    transform: translateX(15px);
                    background-color: $GREEN;

                    &::after {
                        opacity: 1;
                    }
                }
            }

            &:checked:disabled + label {
                #{$toggler}__state {
                    transform: translateX(15px);
                    background-color: $GREY_LIGHT;

                    &::after {
                        opacity: 1;
                        border-color: $GREY_DARK;
                    }
                }
            }

            &:not(:checked):disabled + label {
                #{$toggler}__state {
                    background-color: $GREY_LIGHT;
                }
            }
        }
    }
</style>
