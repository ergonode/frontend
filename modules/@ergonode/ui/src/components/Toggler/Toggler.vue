/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
<template>
    <div class="toggler">
        <input
            :id="associatedLabel"
            :value="value"
            :disabled="disabled"
            :checked="value"
            type="checkbox"
            @input="onValueChange">
        <label :for="associatedLabel">
            <div class="toggler__state-background">
                <div class="toggler__state" />
            </div>
            <span
                v-if="label"
                class="toggler__label"
                v-text="label" />
        </label>
        <slot name="append" />
    </div>
</template>

<script>
import associatedLabelMixin from '@UI/mixins/inputs/associatedLabelMixin';

export default {
    name: 'Toggler',
    mixins: [
        associatedLabelMixin,
    ],
    props: {
        /**
         * Component value
         */
        value: {
            type: Boolean,
            default: false,
        },
        /**
         * Determinate if the component is disabled
         */
        disabled: {
            type: Boolean,
            default: false,
        },
        /**
         * The label is a text caption or description for the component
         */
        label: {
            type: String,
            default: '',
        },
    },
    methods: {
        onValueChange(event) {
            event.preventDefault();

            this.$emit('input', !this.value);
        },
    },
};
</script>

<style lang="scss" scoped>
    .toggler {
        $toggler: &;

        position: relative;
        display: grid;
        grid-template-columns: max-content;
        grid-auto-flow: column;
        column-gap: 8px;
        align-items: center;

        & label {
            position: relative;
            display: grid;
            grid-auto-flow: column;
            column-gap: 8px;
            align-items: center;
        }

        &__state-background {
            display: flex;
            align-items: center;
            width: 30px;
            height: 12px;
            background-color: $GREY_LIGHT;
            border-radius: 8px;
            box-shadow: $ELEVATOR_HOLE;
            cursor: pointer;
        }

        &__state {
            position: relative;
            z-index: $Z_INDEX_LVL_1;
            display: flex;
            width: 16px;
            height: 16px;
            border-radius: 8px;
            background-color: $WHITE;
            box-shadow: $ELEVATOR_2_DP;
            transition: all 0.3s cubic-bezier(0.25, 0.8, 0.5, 1);

            &::after {
                position: absolute;
                top: 5px;
                left: 4px;
                width: 6px;
                height: 3px;
                border-left: 2px solid $WHITE;
                border-bottom: 2px solid $WHITE;
                transform: rotate(-45deg);
                opacity: 0;
                content: "";
            }
        }

        &__label {
            color: $GRAPHITE_DARK;
            font: $FONT_MEDIUM_12_16;
            cursor: pointer;
        }

        & input[type="checkbox"] {
            position: absolute;
            top: 0;
            left: 0;
            opacity: 0;

            &:focus + label, &:not(:disabled):hover + label {
                #{$toggler}__state {
                    box-shadow: $ELEVATOR_HOVER_FOCUS;
                }
            }

            &:checked + label {
                #{$toggler}__state {
                    transform: translateX(15px);
                    background-color: $GREEN;

                    &::after {
                        opacity: 1;
                    }
                }
            }

            &:checked:disabled + label {
                #{$toggler}__state {
                    transform: translateX(15px);
                    background-color: $GREY_LIGHT;

                    &::after {
                        opacity: 1;
                        border-color: $GREY_DARK;
                    }
                }
            }

            &:not(:checked):disabled + label {
                #{$toggler}__state {
                    background-color: $GREY_LIGHT;
                }
            }
        }
    }
</style>
