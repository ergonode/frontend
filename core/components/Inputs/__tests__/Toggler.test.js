/*
 * Copyright Â© Bold Brand Commerce Sp. z o.o. All rights reserved.
 * See LICENSE for license details.
 */
import { shallowMount } from '@vue/test-utils';
import Toggler from '~/core/components/Inputs/Toggler/Toggler';

describe('Inputs/Toggler/Toggler', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(Toggler);
    });

    it('Component is named well', () => {
        expect(typeof Toggler.name).toBe('string');
        expect(Toggler.name).toEqual('Toggler');
        expect(wrapper).toMatchSnapshot();
    });
});
