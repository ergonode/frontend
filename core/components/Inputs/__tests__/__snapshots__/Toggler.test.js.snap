// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputs/Toggler/Toggler Component is named well 1`] = `
<div class="toggler"><input id="input-1" type="checkbox" value="false"> <label for="input-1">
    <div class="toggler__state"></div>
  </label>
  <!---->
</div>
`;
